include/openssl/engine.h

```diff
@@ -19,7 +19,80 @@
 
 # include <openssl/opensslconf.h>
 
-# ifndef OPENSSL_NO_ENGINE
+/*
+ * The logic for stub definitions is driven by these defines:
+ *
+ *  OPENSSL_NO_ENGINE  OPENSSL_ENGINE_STUBS =>   engine.h
+ *       undef              undef              fw decl & info
+ *       undef               def                  stubs
+ *        def               undef                 empty
+ *        def                def                  stubs
+ */
+
+# if !defined(OPENSSL_NO_ENGINE) || defined(OPENSSL_ENGINE_STUBS)
+
+/* OSSL_DEPRECATED is not compatible here, just ignore it */
+#  ifndef OPENSSL_SUPPRESS_DEPRECATED
+#   define ENGINE_STUB_DEPRECATED(msg) OSSL_DEPRECATED_FOR(4.0, msg)
+#  else
+#   define ENGINE_STUB_DEPRECATED(msg)
+#  endif
+
+#  if defined(OPENSSL_ENGINE_STUBS)
+
+#   define ENGINE_STUB_MSG " it is removed. Using stub to avoid compilation error."
+
+#   define ENGINE_FUNC(ret_type, name, params, param_names, default_val) \
+    ENGINE_STUB_DEPRECATED(ENGINE_STUB_MSG) \
+    static inline ret_type name params \
+    { \
+        (void)param_names; /* avoid unused parameter warnings */ \
+        return default_val; /* stub return */ \
+    }
+
+#   define ENGINE_FUNC_VOID(ret_type, name, default_val) \
+    ENGINE_STUB_DEPRECATED(ENGINE_STUB_MSG) \
+    static inline ret_type name(void) \
+    { \
+        return default_val; /* stub return */ \
+    }
+
+#   define ENGINE_VOID_FUNC(name, params, param_names) \
+    ENGINE_STUB_DEPRECATED(ENGINE_STUB_MSG) \
+    static inline void name params \
+    { \
+        (void)param_names; /* avoid unused parameter warnings */ \
+    }
+
+#   define ENGINE_VOID_FUNC_VOID(name) \
```

Comment:
I would rename this to ENGINE_VOID_FUNC_NOARGS so it is clear function receives no arguments and returns void.

Reply:
@Sashan Thanks, it is a proof-of-concept, so I kept stupid names as I had no better idea at the moment :-).
There is also a question is if we want to share these macros in a separate header, then also prefix (ENGINE_) would have to go - so I expect some renaming anyway.